h/tasks/annotations.py

```diff
@@ -0,0 +1,38 @@
+from sqlalchemy import func, update
+
+from h.celery import celery, get_task_logger
+from h.models import Annotation, User
+
+log = get_task_logger(__name__)
+
+
+@celery.task
+def fill_pk_and_user_id(batch_size=1000):
+    """
+    Task to fill the new annotation.pk and annotation.user_id in batches.
+
+    Once most of the existing rows are done we'll make the code changes
+    to keep these up to date, make the column not nullable and remove this task.
+    """
+    # pylint: disable=no-member
+    db = celery.request.db
+
+    annotations = (
+        db.query(Annotation.id.label("annotation_id"), User.id.label("user_id"))
+        .join(
+            User,
+            User.username
+            == func.split_part(func.split_part(Annotation.userid, "@", 1), ":", 2),
+        )
+        .where(Annotation.pk.is_(None))
+        .order_by(Annotation.updated.asc())
+        .limit(batch_size)
+    ).cte("annotations")
```

Comment:
I think you might want to consider [with_for_update(skip_locked=True)](https://docs.sqlalchemy.org/en/14/orm/query.html#sqlalchemy.orm.Query.with_for_update) here. It does an SQL `SELECT FOR UPDATE SKIP LOCKED` query. It'll ensure that if there are two instances of this task running at the same time they don't select the same rows from the table.

This is what the job queue uses: https://github.com/hypothesis/h/blob/41597de80b8e30d7762ee92348d96c6017eea844/h/services/search_index/_queue.py#L229-L236

Reply:
Done :+1: